storage.rb

```diff
@@ -0,0 +1,256 @@
+    options.positional_parameters(query_args)
+
+    res = @cluster.query(query_prep, options)
+    res.rows.each do |row|
+      airport_list.push('airportname' => row['airportname'])
+    end
+
+    { 'context' => ["#{query_type} #{query_prep}"], 'data' => airport_list }
+  end
+
+  def get_flightpaths(from, to, leave)
+    query_type = "N1QL query - scoped to inventory: "
+
+    query_prep = 'SELECT faa as fromAirport FROM `travel-sample`.inventory.airport ' \
+                 'WHERE airportname = $1 ' \
+                 'UNION SELECT faa as toAirport FROM `travel-sample`.inventory.airport ' \
+                 'WHERE airportname = $2'
+    # add first query to context list
+    context = ["#{query_type}\n #{query_prep}"]
+
+    options = Cluster::QueryOptions.new
+    options.positional_parameters([from, to])
+
+    from_airport = ''
+    to_airport = ''
+    res = @cluster.query(query_prep, options)
+    res.rows.each do |row|
+      # Extract the 'fromAirport' and 'toAirport' values
+      from_airport = row['fromAirport'] if row.key? 'fromAirport'
+      to_airport = row['toAirport'] if row.key? 'toAirport'
+    end
+
+    query_routes = 'SELECT a.name, s.flight, s.utc, r.sourceairport, r.destinationairport, r.equipment ' \
+                   'FROM `travel-sample`.inventory.route AS r ' \
+                   'UNNEST r.schedule AS s ' \
+                   'JOIN `travel-sample`.inventory.airline AS a ON KEYS r.airlineid ' \
+                   'WHERE r.sourceairport = $from_faa AND r.destinationairport = $to_faa AND s.day = $day ' \
+                   'ORDER BY a.name ASC '
+    # add second query to context list
+    context.push("#{query_type}\n #{query_routes}")
+
+    options = Cluster::QueryOptions.new
+    options.named_parameters({ 'from_faa' => from_airport, 'to_faa' => to_airport, 'day' => convdate(leave) })
+
+    route_list = []
+    res = @cluster.query(query_routes, options)
+    res.rows.each do |row|
+      row['flighttime'] = (rand() * 8000).ceil
+      row['price'] = (row['flighttime'].to_f / 8 * 100 / 100).ceil(2)
+      route_list.push(row)
+    end
+
+    { 'context' => context, 'data' => route_list }
+  end
+
+  def get_hotels(description, location)
+    search = Cluster::SearchQuery
+    search_cols = ['address', 'city', 'state', 'country', 'name', 'description']
+    qp = search.conjuncts
+
+    # Fallback in case we want to search for any hotels.
+    # If this is not present, couchbase will error with `ArgumentError: compound conjunction query must have sub-queries`
+    qp.and_also(search::TermQuery.new('hotel'))
+
+    if location != "*" and location != ""
+      qp.and_also(
+        search.disjuncts(
+          search.match_phrase(location) { |q| q.field = "country" },
+          search.match_phrase(location) { |q| q.field = "city" },
+          search.match_phrase(location) { |q| q.field = "state" },
+          search.match_phrase(location) { |q| q.field = "address" },
```

Comment:
I've glazed over the FTS syntax in other apps, because I haven't learnt much about it, but for some reason just now I'm realising how odd this construction is, with the lambda on the query object to construct it.

Reply:
Yeah, I still need to understand FTS syntax a bit better tbh. This approach was something @avsej suggested as I wasn't sure how to do it in Ruby in an idiomatic way. 🤷‍♀️

Reply:
Yeah! I'm sure it's fine, I just stared it and went "huh?" this time. Will add FTS to list of things to learn properly over the summer...